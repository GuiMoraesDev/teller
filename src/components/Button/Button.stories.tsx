import React from 'react'

import { ComponentMeta, ComponentStory } from '@storybook/react'

import Button from '.'

const Storie = {
  title: 'Components/Button',
  component: Button
} as ComponentMeta<typeof Button>

export default Storie

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />

export const PrimaryVariant = Template.bind({})

PrimaryVariant.args = {
  type: 'button',
  label: 'Random Button',
  fullWidth: false,
  dimension: 'sm',
  rounded: 'sm',
  variant: 'primary'
}

export const SecondaryVariant = Template.bind({})

SecondaryVariant.args = {
  type: 'button',
  label: 'Random Button',
  fullWidth: false,
  dimension: 'sm',
  rounded: 'sm',
  variant: 'secondary'
}

export const NeutralVariant = Template.bind({})

NeutralVariant.args = {
  type: 'button',
  label: 'Random Button',
  fullWidth: false,
  dimension: 'sm',
  rounded: 'sm',
  variant: 'neutral'
}

export const OutlineVariant = Template.bind({})

OutlineVariant.args = {
  type: 'button',
  label: 'Random Button',
  fullWidth: false,
  dimension: 'sm',
  rounded: 'sm',
  variant: 'outline'
}

export const AlertVariant = Template.bind({})

AlertVariant.args = {
  type: 'button',
  label: 'Random Button',
  fullWidth: false,
  dimension: 'sm',
  rounded: 'sm',
  variant: 'alert'
}

export const DisabledVariant = Template.bind({})

DisabledVariant.args = {
  type: 'button',
  label: 'Disabled Button',
  fullWidth: false,
  disabled: true,
  dimension: 'sm',
  rounded: 'sm',
  variant: 'primary'
}

export const RoundedVariant = Template.bind({})

RoundedVariant.args = {
  type: 'button',
  label: 'Random Button',
  fullWidth: false,
  dimension: 'sm',
  rounded: 'full',
  variant: 'primary'
}

export const IconBasic = Template.bind({})

IconBasic.args = {
  type: 'button',
  iconLeft: 'Plus',
  fullWidth: false,
  dimension: 'sm',
  rounded: 'sm',
  variant: 'primary'
}

export const LabelIconLeftBasic = Template.bind({})

LabelIconLeftBasic.args = {
  type: 'button',
  label: 'Random Button',
  iconLeft: 'Leaf',
  fullWidth: false,
  dimension: 'sm',
  rounded: 'sm',
  variant: 'primary'
}

export const LabelIconRightBasic = Template.bind({})

LabelIconRightBasic.args = {
  type: 'button',
  label: 'Random Button',
  iconRight: 'Plus',
  fullWidth: false,
  dimension: 'sm',
  rounded: 'sm',
  variant: 'primary'
}

export const LabelIconLeftRightBasic = Template.bind({})

LabelIconLeftRightBasic.args = {
  type: 'button',
  label: 'Random Button',
  iconLeft: 'Plus',
  iconRight: 'Minus',
  fullWidth: false,
  dimension: 'sm',
  rounded: 'sm',
  variant: 'primary'
}
