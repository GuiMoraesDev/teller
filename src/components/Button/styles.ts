import styled, { css, keyframes } from 'styled-components';

import { ButtonDefaultPropsThatMakeStyles } from '.';

const rotate = keyframes`
	from {
		transform: rotate(0deg);
	}
	to {
		transform: rotate(360deg);
	}
`;

export const Button = styled.button<ButtonDefaultPropsThatMakeStyles>`
	position: relative;

	display: flex;
	align-items: center;
	justify-content: center;

	width: ${({ fullWidth }) => (fullWidth ? '100%' : 'auto')};
	height: ${({ theme, dimension }) =>
		theme.sizes.dimensions.height[dimension || 'md']};

	border: ${({ theme }) => theme.borders.solid};
	border-radius: ${({ theme, rounded }) => theme.rounded[rounded || 'none']};

	margin: 0;
	padding: 0
		${({ theme, dimension }) =>
			theme.sizes.dimensions.spacing[dimension || 'md']};
	gap: ${({ theme }) => theme.sizes.dimensions.spacing.sm};

	transition: background-color ${({ theme }) => theme.transition.fast},
		color ${({ theme }) => theme.transition.fast};

	cursor: pointer;

	&:hover {
		background-color: ${({ theme }) => theme.themeColors.brandPrimaryLight};
	}

	svg {
		color: inherit;
	}

	${({ variant, theme }) => {
		if (variant === 'primary')
			return css`
				color: ${theme.themeColors.textInverted};

				background-color: ${theme.themeColors.brandPrimary};

				border-color: ${theme.themeColors.brandPrimaryDark};

				&:hover {
					background-color: ${theme.themeColors.brandPrimaryDark};
				}
			`;

		if (variant === 'secondary')
			return css`
				color: ${theme.themeColors.textInverted};

				background-color: ${theme.themeColors.canvasInvertedLight};

				border-color: ${theme.themeColors.canvasInverted};

				&:hover {
					background-color: ${theme.themeColors.canvasInverted};
				}
			`;

		if (variant === 'outline')
			return css`
				color: ${theme.themeColors.text};

				background-color: ${theme.themeColors.canvas};

				border-color: ${theme.themeColors.canvasInvertedLight};

				&:hover {
					color: ${theme.themeColors.textInverted};

					background-color: ${theme.themeColors.canvasInverted};
				}
			`;

		if (variant === 'neutral')
			return css`
				color: ${theme.themeColors.text};

				background-color: ${theme.colors.transparent};

				border-color: transparent;

				&:hover {
					text-decoration: underline;

					color: ${theme.themeColors.text};

					background-color: ${theme.colors.transparent};
				}
			`;

		if (variant === 'alert')
			return css`
				color: ${theme.themeColors.text};

				background-color: ${theme.colors.danger[300]};

				border-color: ${theme.colors.danger[500]};

				&:hover {
					background-color: ${theme.colors.danger[500]};
				}
			`;
	}}

	${({ theme, isDisabled }) => {
		if (isDisabled)
			return css`
				color: ${theme.colors.neutrals[500]};

				background-color: ${theme.colors.neutrals[200]};

				border: ${theme.borders.none};

				&:hover {
					color: ${theme.colors.neutrals[500]};

					background-color: ${theme.colors.neutrals[200]};
				}
			`;
	}}

	.spin {
		animation-name: ${rotate};
		animation-duration: 2s;
		animation-iteration-count: infinite;
		animation-timing-function: linear;
	}
`;
