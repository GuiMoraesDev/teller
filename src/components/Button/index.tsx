import { ButtonHTMLAttributes, forwardRef, useCallback } from 'react';

import { useRouter } from 'next/router';

import { DimensionsProps } from '@types';
import { Spinner } from 'phosphor-react';
import { DefaultTheme } from 'styled-components';

import * as Styles from './styles';

export interface ButtonDefaultPropsThatMakeStyles {
	fullWidth?: boolean;
	variant?: 'primary' | 'secondary' | 'outline' | 'neutral' | 'alert';
	dimension?: keyof DimensionsProps;
	rounded?: keyof DefaultTheme['rounded'];
	isDisabled: boolean;
	IconLeft?: JSX.Element;
	IconRight?: JSX.Element;
	isLoading?: boolean;
	isCapitalize?: boolean;
}

export interface ButtonDefaultProps
	extends Omit<ButtonDefaultPropsThatMakeStyles, 'isDisabled'> {
	label?: string;
	disabled?: boolean;
	href?: string;
	onClick?(event: React.MouseEvent<HTMLButtonElement>): void;
}

export type ButtonProps = ButtonDefaultProps &
	ButtonHTMLAttributes<HTMLButtonElement>;

const Button: React.ForwardRefRenderFunction<HTMLButtonElement, ButtonProps> = (
	{
		label,
		variant = 'primary',
		IconLeft,
		IconRight,
		fullWidth = false,
		dimension = 'md',
		rounded = 'sm',
		href,
		isLoading,
		...props
	},
	ref
): JSX.Element => {
	const navigate = useRouter();

	const handleClick = useCallback(
		(event: React.MouseEvent<HTMLButtonElement>) => {
			if (href) {
				return navigate.push(href);
			}

			return props.onClick?.(event);
		},
		[href, navigate, props]
	);

	return (
		<Styles.Button
			fullWidth={fullWidth}
			dimension={dimension}
			rounded={rounded}
			variant={variant}
			onClick={handleClick}
			isDisabled={Boolean(props.disabled)}
			{...props}
			ref={ref}
		>
			{IconLeft}

			{label}

			{isLoading ? <Spinner className="spin" /> : IconRight}
		</Styles.Button>
	);
};

export default forwardRef(Button);
