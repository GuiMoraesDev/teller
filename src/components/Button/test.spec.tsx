import React from "react";

import { render, fireEvent } from "tests/test.utils";

import Button from "./index";

describe("Button component", () => {
  it("should render correctly", () => {
    const { container } = render(<Button type="button" />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render visible", () => {
    const { container } = render(<Button type="button" />);

    expect(container.firstChild).toBeVisible();
  });

  it("should call onClick function", () => {
    const mockedOnClick = jest.fn();

    const { getByText } = render(
      <Button type="button" label="jest button" onClick={mockedOnClick} />
    );

    fireEvent.click(getByText(/jest button/i));

    expect(mockedOnClick).toBeCalled();
  });

  it("should be able to submit a form", () => {
    const mockedOnSubmit = jest.fn((e) => e.preventDefault());

    const { getByText } = render(
      <form onSubmit={mockedOnSubmit}>
        <Button type="submit" label="jest button" />
      </form>
    );

    fireEvent.click(getByText(/jest button/i));

    expect(mockedOnSubmit).toBeCalled();
  });
});
