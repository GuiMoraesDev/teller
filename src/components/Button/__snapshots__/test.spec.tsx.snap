// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component should render correctly 1`] = `
.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: auto;
  height: 4.4rem;
  border: 0.1rem solid #C9CAC8;
  border-radius: 0.4rem;
  margin: 0;
  padding: 0 1.4rem;
  gap: 1.2rem;
  text-transform: none;
  -webkit-transition: background-color 0.2s, color 0.2s;
  transition: background-color 0.2s, color 0.2s;
  cursor: pointer;
  color: #FDFDFD;
  background-color: #63989F;
  border-color: #4F7A7F;
}

.c0:hover {
  background-color: #82ADB2;
}

.c0 svg {
  color: inherit;
}

.c0:hover {
  background-color: #4F7A7F;
}

.c0 .spin {
  -webkit-animation-name: ieAXeN;
  animation-name: ieAXeN;
  -webkit-animation-duration: 2s;
  animation-duration: 2s;
  -webkit-animation-iteration-count: infinite;
  animation-iteration-count: infinite;
  -webkit-animation-timing-function: linear;
  animation-timing-function: linear;
}

<button
  class="c0"
  type="button"
/>
`;
